### PR TITLE
Fix hangs when worker main pool has failures.

### PR DESCRIPTION
**Pull request by fyrestone**
## What do these changes do?

- Improve fault injection, support fault position and fault context.
- New `patch_cls` and `patch_super` in mars.tests.core
- Fix hangs when worker main pool has failures, e.g.
  - Unhandled exceptions in SubtaskExecutionActor.
  - Worker main pool crashes.

### CODE REVIEW

File: mars/services/scheduling/worker/execution.py [hunk 1]
```diff
@@ -32,6 +32,12 @@
 from ...meta import MetaAPI
 from ...storage import StorageAPI
 from ...subtask import Subtask, SubtaskAPI, SubtaskResult, SubtaskStatus
+from ...tests.fault_injection_manager import (
```
Comment (wjsi) [inline mars/services/scheduling/worker/execution.py:35]: Test modules shall never be used in other part of the code.

Comment (fyrestone) [reply mars/services/scheduling/worker/execution.py:35]: Is there better ways to inject fault to test the code?

Comment (wjsi) [reply mars/services/scheduling/worker/execution.py:35]: > Is there better ways to inject fault to test the code?

I think it is better to use `patch` module. Some work is needed to pass method patches to different processes.

Comment (fyrestone) [reply mars/services/scheduling/worker/execution.py:35]: > > Is there better ways to inject fault to test the code?
> 
> I think it is better to use `patch` module. Some work is needed to pass method patches to different processes.

Can you give more information about the `patch` module?

**Review by wjsi — APPROVED** [review]
LGTM

**Review by qinxuye — APPROVED** [review]
LGTM

**Pull request merged**